### DISCUSSION TITLE
Creating an Input Schema for Multiple DataFrames in MLflow

### DISCUSSION THREAD

**Discussion by ecram** [mlflow/mlflow#12723] [Q&A]
Hi everyone,

I am working with MLflow version 2.5.0 and need to create an input_schema for my model. My data schema is divided into several DataFrames, for example:

{
"dataframe_split": {
     "columns": ["ClientGuid", "Instance", "TypeScore", ...],
     "data": [["26c5f035-cf57-4cb7-b67c-55da528ce3e0", "1", "POST", ...]]
},
"dataframe_loan_split": {
     "columns": ["System", "Entity", "Credit", ...],
     "data": [["R", "Entity1", "M1 - MIC. IND. DEB. GAR.", ...]]
},
...
}

Could someone provide an example of how to define an input_schema using MLflow for this type of structure? I am looking for a way to specify all the DataFrames within the input schema.

Thanks for your help.

**Comment by B-Step62** [comment]
@ecram Currently MLflow model does not support passing multiple Pandas DataFrame with different schema as model input. Would it be possible to combine your input, or use different format such as a list of dictionary? With the introduction of Arrays signature in MLflow 2.10.0, MLflow can handle complex objects like nested dictionary as input schema. Also curious to know what kind of input your model expects.

**Reply by SebastianScherer88** [reply]
i'm not the OP, but have the same requirement. Essentially, this stems from the attempt to include the model specific preprocessing in the mlflow custom model predict method. Preprocessing might take multiple data sources at various levels: 
- lender features (i.e.1 row=1lender), 
- lender's transaction features (i.e. 1row=1 financial transaction), etc
which get processed into a dataset where 1row= 1model observation (i.e. the ML model feature dataset).

it would be nice for mlflow's schema inference and validation logic to support this in its most common implementation (imo): a dictionary of dataframes. you already support dictionaries of arrays, so this would be a super useful extension of that.

**Reply by ecram** [reply]
Hi,  In our case, we received the data in a JSON format where each key-value pair, corresponded to a different DataFrame.

By organizing the inputs this way — with each part of the JSON mapping to a DataFrame — we were able to handle multiple data sources (lender-level features, transaction-level features, etc.) cleanly inside the predict method of our MLflow custom model. This approach effectively solved the preprocessing integration issue you mentioned.

From this format, we use the JSON library and obtain the data frames separated into key pairs individually for processing.

It's always recommended to obtain a single data frame object for processing, but this may help you.
